File: Python/programmers/kakao/level4/가사_검색.py
def search(head, query):
    cur = head
    res = 0
    for char in query:
        if char == "?":
            return res
        else:
            if char not in cur[1]:
                return 0
            res = cur[1][char][0]
            cur = cur[1][char]
    return res


def solution(words, queries):
    answer = []
    prefix, suffix = {}, {}

    for word in words:
        wlen = len(word)
        if wlen not in prefix:
            prefix[wlen], suffix[wlen] = [0, {}], [0, {}]
        for dictionary, string in [prefix, word], [suffix, word[::-1]]:
            cur = dictionary[wlen]
            cur[0] += 1
            for char in string:
                if char not in cur[1]:
                    cur[1][char] = [0, {}]
                cur[1][char][0] += 1
                cur = cur[1][char]
    
    for query in queries:
        qlen = len(query)
        if qlen not in prefix:
            answer.append(0)
        elif query[0] == "?" and query[-1] == "?":
            answer.append(prefix[qlen][0])
        elif query[-1] == "?":
            answer.append(search(prefix[qlen], query))
        elif query[0] == "?":
            answer.append(search(suffix[qlen], query[::-1]))

    return answer


words = ["frodo", "front", "frost", "frozen", "frame", "kakao"]
queries = ["fro??", "????o", "fr???", "fro???", "pro?"]
print(solution(words, queries)) # [3, 2, 4, 1, 0]

File: Java/programmers/kakao/level4/가사검색.java
package CodeTest.Java.programmers.kakao.level4;

import java.util.*;

class Trie {
    Map<Integer, Integer> lenMap = new HashMap<>();
    Trie[] child = new Trie[26];

    void insert(String st) {
        Trie node = this;
        int len = st.length();
        lenMap.put(len, lenMap.getOrDefault(len, 0)+1);

        for (char ch : st.toCharArray()) {
            int idx = ch-'a';
            if (node.child[idx] == null) node.child[idx] = new Trie();

            node = node.child[idx];
            node.lenMap.put(len, node.lenMap.getOrDefault(len, 0)+1);
        }
    }

    int find(String st, int i) {
        if (st.charAt(i) == '?') return lenMap.getOrDefault(st.length(), 0);

        int idx = st.charAt(i)-'a';
        return child[idx] == null ? 0 : child[idx].find(st, i+1);
    }
}

public class 가사검색 {
    public int[] solution(String[] words, String[] queries) {
        Trie front = new Trie(), back = new Trie();

        for (String word : words) {
            front.insert(word);
            back.insert(reverse(word));
        }

        int[] answer = new int[queries.length];

        for (int i=0; i<queries.length; i++) {
            String query = queries[i];
            if (query.charAt(0) != '?') answer[i] = front.find(query, 0);
            else answer[i] = back.find(reverse(query), 0);
        }

        return answer;
    }

    private static String reverse(String st) {
        return new StringBuilder(st).reverse().toString();
    }
}
